RecoParticleFlow/PFRecHitProducer/plugins/alpaka/CalorimeterDefinitions.h

```diff
@@ -0,0 +1,207 @@
+#include "RecoParticleFlow/PFRecHitProducer/interface/alpaka/PFRecHitParamsDeviceCollection.h"
+#include "RecoParticleFlow/PFRecHitProducer/interface/alpaka/PFRecHitTopologyDeviceCollection.h"
+
+// This file defines two structs:
+// 1) ALPAKA_ACCELERATOR_NAMESPACE::particleFlowRecHitProducer::HCAL
+// 2) ALPAKA_ACCELERATOR_NAMESPACE::particleFlowRecHitProducer::ECAL
+// These are used as template arguments of the PFRecHitSoAProducer class and
+// related classes. This allows to specialise behaviour for the two calorimeter
+// types.
+namespace ALPAKA_ACCELERATOR_NAMESPACE::particleFlowRecHitProducer {
+
+  // Get subdetector encoded in detId to narrow the range of reference table values to search
+  constexpr inline uint32_t getSubdet(uint32_t detId) { return DetId(detId).subdetId(); }
+
+  struct HCAL {
+    using CaloRecHitType = HBHERecHit;
+    using CaloRecHitSoATypeHost = reco::CaloRecHitHostCollection;
+    using CaloRecHitSoATypeDevice = reco::CaloRecHitDeviceCollection;
+    using ParameterType = reco::PFRecHitHCALParamsDeviceCollection;
+    using ParameterRecordType = PFRecHitHCALParamsRecord;
+    using TopologyTypeHost = reco::PFRecHitHCALTopologyHostCollection;
+    using TopologyTypeDevice = reco::PFRecHitHCALTopologyDeviceCollection;
+    using TopologyRecordType = PFRecHitHCALTopologyRecord;
+
+    static constexpr DetId::Detector kDetectorId = DetId::Detector::Hcal;
+    static constexpr int kSubdetectorBarrelId = HcalSubdetector::HcalBarrel;
+    static constexpr int kSubdetectorEndcapId = HcalSubdetector::HcalEndcap;
+
+    static constexpr uint32_t kMaxDepthHB = 4;
+    static constexpr uint32_t kMaxDepthHE = 7;
+    static constexpr uint32_t kFirstHBRing = 1;
+    static constexpr uint32_t kLastHBRing = 16;
+    static constexpr uint32_t kFirstHERing = 16;
+    static constexpr uint32_t kLastHERing = 29;
+    static constexpr uint32_t kMaxIPhi = 72;
+    static constexpr uint32_t kSizeBarrel = kMaxDepthHB * (kLastHBRing - kFirstHBRing + 1) * kMaxIPhi * 2;
+    static constexpr uint32_t kSizeEndcap = kMaxDepthHE * (kLastHERing - kFirstHERing + 1) * kMaxIPhi * 2;
+    static constexpr uint32_t kSize = kSizeBarrel + kSizeEndcap;  // maximum possible HCAL denseId (=23328)
+
+    static constexpr bool detIdInRange(uint32_t detId) {
+      return detId != 0 && DetId(detId).det() == DetId::Detector::Hcal &&
+             (getSubdet(detId) == HcalSubdetector::HcalBarrel || getSubdet(detId) == HcalSubdetector::HcalEndcap);
+    }
+
+    static constexpr uint32_t getDepth(uint32_t detId) { return HcalDetId(detId).depth(); }
+    static constexpr uint32_t getIetaAbs(uint32_t detId) { return HcalDetId(detId).ietaAbs(); }
+    static constexpr uint32_t getIphi(uint32_t detId) { return HcalDetId(detId).iphi(); }
+    static constexpr int getZside(uint32_t detId) { return HcalDetId(detId).zside(); }
+
+    // https://cmssdt.cern.ch/lxr/source/Geometry/CaloTopology/src/HcalTopology.cc#1170
```

Comment:
Could this copy-paste of logic be avoided in some way? (possibly for a subsequent PR, and also elsewhere in this file)